fvm/evm/emulator/database/database.go

```diff
@@ -0,0 +1,341 @@
+	atreemap           *atree.OrderedMap
+	// Ramtin: other database implementations for EVM uses a lock
+	// to protect the storage against concurrent operations
+	// though one might do more research to see if we need
+	// these type of locking if the underlying structure (atree)
+	// has such protections or if EVM really needs it
+	lock sync.RWMutex
+}
+
+// NewDatabase returns a wrapped map that implements all the required database interface methods.
+func NewDatabase(led atree.Ledger, flowEVMRootAddress flow.Address) (*Database, error) {
+	baseStorage := atree.NewLedgerBaseStorage(led)
+
+	storage, err := NewPersistentSlabStorage(baseStorage)
+	if err != nil {
+		return nil, handleError(err)
+	}
+
+	db := &Database{
+		led:                led,
+		flowEVMRootAddress: flowEVMRootAddress,
+		storage:            storage,
+	}
+
+	err = db.retrieveOrCreateMapRoot()
+	if err != nil {
+		return nil, err
+	}
+	return db, nil
+}
+
+func (db *Database) retrieveOrCreateMapRoot() error {
+	rootIDBytes, err := db.led.GetValue(db.flowEVMRootAddress.Bytes(), []byte(FlowEVMRootSlabKey))
+	if err != nil {
+		return handleError(err)
+	}
+
+	var m *atree.OrderedMap
+	if len(rootIDBytes) == 0 {
+		m, err = atree.NewMap(db.storage, atree.Address(db.flowEVMRootAddress), atree.NewDefaultDigesterBuilder(), typeInfo{})
+		if err != nil {
+			return handleError(err)
+		}
+	} else {
+		storageID, err := atree.NewStorageIDFromRawBytes(rootIDBytes)
+		if err != nil {
+			return handleError(err)
+		}
+		m, err = atree.NewMapWithRootID(db.storage, storageID, atree.NewDefaultDigesterBuilder())
+		if err != nil {
+			return handleError(err)
+		}
+	}
+	db.atreemap = m
+	return nil
+}
+
+func (db *Database) storeMapRoot() error {
+	rootIDBytes := make([]byte, StorageIDSize)
+	_, err := db.atreemap.StorageID().ToRawBytes(rootIDBytes)
+	if err != nil {
+		return handleError(err)
+	}
+	err = db.led.SetValue(db.flowEVMRootAddress.Bytes(), []byte(FlowEVMRootSlabKey), rootIDBytes[:])
+	if err != nil {
+		return handleError(err)
+	}
+	return nil
+}
+
+// Get retrieves the given key if it's present in the key-value store.
+func (db *Database) Get(key []byte) ([]byte, error) {
+	db.lock.RLock()
+	defer db.lock.RUnlock()
+
+	return db.get(key)
+}
+
+func (db *Database) get(key []byte) ([]byte, error) {
+	data, err := db.atreemap.Get(compare, hashInputProvider, NewByteStringValue(key))
+	if err != nil {
+		return nil, handleError(err)
+	}
+
+	v, err := data.StoredValue(db.atreemap.Storage)
+	if err != nil {
+		return nil, handleError(err)
+	}
+
+	return v.(ByteStringValue).Bytes(), nil
+}
+
+// Put inserts the given value into the key-value store.
+func (db *Database) Put(key []byte, value []byte) error {
+	db.lock.Lock()
+	defer db.lock.Unlock()
+	return db.put(key, value)
+}
+
+func (db *Database) put(key []byte, value []byte) error {
+	_, err := db.atreemap.Set(compare, hashInputProvider, NewByteStringValue(key), NewByteStringValue(value))
+	return handleError(err)
```

Comment:
Atree `OrderedMap.Set` returns overwritten storable if key exists.  We probably want to remove overwritten `storable` from storage if `storable` is `StorageIDStorable`.  

Maybe something like this:
```suggestion
	existingValueStorable, err := db.atreemap.Set(compare, hashInputProvider, NewByteStringValue(key), NewByteStringValue(value))
	if err != nil {
		return handleError(err)
	}
	
	if id, ok := existingValueStorable.(atree.StorageIDStorable); ok {
		// NOTE: deep remove isn't necessary because value is ByteStringValue (not container)
		err := db.storage.Remove(id)
	 	if err != nil {
	  		return handleError(err)
		}
	}
	
	return nil
```